zebra-chain/src/sprout/joinsplit.rs

```diff
@@ -89,6 +119,55 @@ impl<P: ZkSnarkProof> JoinSplit<P> {
 
         (vpub_new - vpub_old).expect("subtraction of two valid amounts is a valid NegativeAllowed")
     }
+
+    /// Encodes the primary inputs for the proof statement as Bls12_381 base
+    /// field elements, to match bellman::groth16::verify_proof.
+    ///
+    /// NB: jubjub::Fq is a type alias for bls12_381::Scalar.
+    ///
+    /// `joinsplit_pub_key`: the JoinSplit public validation key for this JoinSplit, from
+    /// the transaction. (All JoinSplits in a transaction share the same validation key.)
+    ///
+    /// This is not yet officially documented; see the reference implementation:
+    /// https://github.com/zcash/librustzcash/blob/0ec7f97c976d55e1a194a37b27f247e8887fca1d/zcash_proofs/src/sprout.rs#L152-L166
+    pub fn primary_inputs(
+        &self,
+        joinsplit_pub_key: &ed25519::VerificationKeyBytes,
+    ) -> Vec<jubjub::Fq> {
+        let rt: [u8; 32] = self.anchor.into();
+        let mac1: [u8; 32] = (&self.vmacs[0]).into();
+        let mac2: [u8; 32] = (&self.vmacs[1]).into();
+        let nf1: [u8; 32] = (&self.nullifiers[0]).into();
+        let nf2: [u8; 32] = (&self.nullifiers[1]).into();
+        let cm1: [u8; 32] = (&self.commitments[0]).into();
+        let cm2: [u8; 32] = (&self.commitments[1]).into();
+        let vpub_old = self.vpub_old.to_bytes();
+        let vpub_new = self.vpub_new.to_bytes();
+
+        let h_sig = h_sig(
+            &self.random_seed[..],
+            &nf1[..],
+            &nf2[..],
+            joinsplit_pub_key.as_ref(),
+        );
+
+        // Prepare the public input for the verifier
+        let mut public_input = Vec::with_capacity((32 * 8) + (8 * 2));
+        public_input.extend(rt);
+        public_input.extend(h_sig);
+        public_input.extend(nf1);
+        public_input.extend(mac1);
+        public_input.extend(nf2);
+        public_input.extend(mac2);
+        public_input.extend(cm1);
+        public_input.extend(cm2);
+        public_input.extend(vpub_old);
+        public_input.extend(vpub_new);
+
+        let public_input = multipack::bytes_to_bits(&public_input);
+
+        multipack::compute_multipacking(&public_input)
+    }
```

Comment:
Proposal: since we already depend on `bellman` in `zebra-consensus`, we could implement this method as a new trait on `JoinSplit` in that crate, so as not to pull in `bellman` into `zebra-chain` just to implement this here (also because this method is not used within `zebra-chain`) ?

Reply:
That makes sense, done in 2051e9ea

Reply:
💯